### DISCUSSION TITLE
Defining database columns, but not using the CodingKeys enum

### DISCUSSION THREAD

**Discussion by hell001** [groue/GRDB.swift#1098] [Q&A]
### What did you do?

I have an API that has a `title` key, but in my code I'd like to have it named `name`. I create a `CodingKeys` enum so that my app can properly read the JSON and also use the variable name I choose. See a truncated version of the class below:

```
class A: Codable, Identifiable {
    var name: String
    
    enum CodingKeys: String, CodingKey {
        case name = "title"
    }
}
```

As per [Define Record Requests](https://github.com/groue/GRDB.swift/blob/master/Documentation/GoodPracticesForDesigningRecordTypes.md#define-record-requests), I created a `Columns` enum as follows below. However, instead of using the CodingKeys value `CodingKeys.name`, I used the String of the variable name itself, which is what I desire the column name to be.

```
extension A: FetchableRecord, MutablePersistableRecord {
    enum Columns {
        static let name = Column("name")
    }
}
```

So to clarify, I still would like to use `CodingKeys` to go from `title` from API response to `name` in Swift code, but still want to use `name` as the GRDB column name, and not take it from `CodingKeys`, which would be `title`.

For extra clarification, the errors occur on my `aInstance.insert(db)` calls.  `table a has no column named title`

### What did you expect to happen?

The column name is `name`.

### What happened instead?

The column name is `title`.

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:**
**Installation method:** SPM
**Xcode version:**
**Swift version:** 5
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 11.0

**Comment by groue** [comment]
Hello @hell001,

Please show how you define this name/title `Column` value.

**Comment by hell001** [comment]
Hello @groue,

Assuming you are not referring to any of the code in the original post,

```
class A: Codable, Identifiable {
    var name: String

    enum CodingKeys: String, CodingKey {
        case name = "title"
    }
}
```

**Comment by groue** [comment]
> Assuming you are not referring to any of the code in the original post,

Whenever someone asks for support, I _always_ refer to the actual code that creates the actual problem. Everything else is just noise.

Your answer does not contain any `Column` value. However you explicitly ask that the `name` property is encoded in a `title` JSON key or database column.

I'm not sure I can help with the available information. If you want help from this discussion, please provide a [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example).

**Comment by hell001** [comment]
does the following code from my original post not include a column value? (quoted below for your convenience)

```
//1
enum Columns {
        static let name = Column("name")
}

//2
enum Columns: String, ColumnExpression {
        case name
}
```

**Reply by groue** [reply]
> please provide a [Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example).

**Reply by hell001** [reply]
@groue Original question has been updated, take a gander. I <3 the quick response times

**Reply by groue** [reply]
Thanks. So you rely on a single `Codable` implementation to use sometimes a `title` key (in the database) and sometimes a `name` key (in JSON). There lies your conflict.

There are several ways to solve it, and the simplest may be to provide an explicit implementation of `MutablePersistableRecord.encode`. See documentation for more details: https://github.com/groue/GRDB.swift/blob/master/README.md#persistablerecord-protocol

**Reply by hell001** [reply]
@groue This is definitely putting me on the right path. After implementing this protocol function `encode` the database is successfully written to, and I can confirm correctly as when I open the SQLite file, everything looks as intended as far as column names. My next question is, is there some equivalent decoding function that needs to be implemented? After expanding on the project a bit, I seem to get an error of:

```
Fatal error: 'try!' expression unexpectedly raised an error: column not found: "title" - row: [id:"1" name:"Name"], sql: `SELECT "a".* FROM "a" JOIN "ab" ON ("ab"."aId" = "a"."id") AND ("ab"."bId" = ?)`, arguments: ["1"]
```
Why is GRDB trying to select `title`? I bring in another class `B` because if I were to simply call `A.fetchAll(db)`, this works with no problem, which confuses me. I Ctrl+F the entire project and the only mention of `title` is in `CodingKeys`.

To clarify, I am receiving the above error when calling `b.aList.fetchAll(db)`

See all code below:

```
class A: Codable, Identifiable {
    var id: String
    var name: String
    
    enum CodingKeys: String, CodingKey {
        case id
        case name = "title"
    }
}

extension A: FetchableRecord, MutablePersistableRecord {}

extension A {
    static let ab = hasMany(AB.self)
    static let bList = hasMany(B, through: ab, using: AB.b)
}

extension A {
    var bList: QueryInterfaceRequest<B> {
        request(for: A.bList)
    }
}

extension A: EncodableRecord {
    func encode(to container: inout PersistenceContainer) {
        container["name"] = name
    }
}

class B: Codable, Identifiable {
    var id: String
}

extension B: FetchableRecord, MutablePersistableRecord {}

extension B {
    static let ab = hasMany(AB.self)
    static let aList = hasMany(A.self, through: ab, using: AB.a)
}

extension B {
    var aList: QueryInterfaceRequest<A> {
        request(for: B.aList)
    }
}

class AB: Codable {
    var aId: String
    var bId: String
    
    init(aId: String, bId: String) {
        self.aId = aId
        self.bId = bId
    }
}

extension AB: FetchableRecord, PersistableRecord {}

extension AB {
    static let a = belongsTo(A.self)
    static let b = belongsTo(B.self)
}
```

Let me know if I should turn this into an issue post. I figured I was missing something tho

**Reply by groue** [reply]
Oh, so you did not read a single line of the doc, and you expect me to digest it for you. This is a very wrong expectation.

[Records](https://github.com/groue/GRDB.swift/blob/master/README.md#records)

**Reply by hell001** [reply]
@groue If I “did not read a single line of the doc” how would I know what this sharp tool even is? Let’s not exaggerate. What I am saying is that even _after_ reading the docs, I am failing to see the part that addresses my previous comment. And as I stated I could simply be missing it or even misinterpreting when going through, in which case I ask for help. After reading the documentation and trying to implement something myself, and after that still I cannot do it, what else should I do that isn’t asking for help with a very detailed question, providing “a Minimal, Reproducible Example.“ ?

**Reply by hell001** [reply]
@groue Following up...

**Reply by groue** [reply]
@hell001, you have recently damaged two other discussions, #1055 and #1049. I have repaired the one that can raise the interest of other users. Please do not do that again.
